Qt: custom playlist model and playlist view

CozmoP opened the pull request:
## Description

Implement custom playlist model and grid view.
Only load images when they become visible and cache them.
Add option to change thumbnail type displayed in grid view.
Add option to change thumbnail cache limit.

## Reviewers

@bparker06

twinaphex commented:
Thanks, I'll let @bparker06 review this since he made the Qt desktop UI. Thanks again for your contribution.

Tatsuya79 commented:
Testing it for now, pretty cool and fast!
I'm just noticing a crash in Grid view when having a search active and switching playlists.
It's a bit random, seems to happen faster when searching with at least 3 letters.

win7 x64, compiled with MSYS2 GCC7.3

Tatsuya79 commented:
The item selection highlight sometimes stop showing in grid view.
The blue highlight in that pic:
![untitled-1](https://user-images.githubusercontent.com/6266038/49903865-cb8efe00-fe68-11e8-9a72-086e9f17fe09.png)

It happens in long playlists after moving back and forth for a while.
There was a point where I couldn't select an item or launch it at all.

bparker06 commented:
I also get a bunch of TSAN warnings but not sure if they are false positive or not:
```
==================
WARNING: ThreadSanitizer: data race (pid=16366)
  Read of size 8 at 0x7ba000058198 by main thread:
    #0 memmove /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:720 (libtsan.so.0+0x3295e)
    #1 QCoreApplication::postEvent(QObject*, QEvent*, int) <null> (libQt5Core.so.5+0x275498)
    #2 GridView::refresh() ui/drivers/qt/gridview.cpp:226 (retroarch+0x2447ca)
    #3 GridView::reset() ui/drivers/qt/gridview.cpp:257 (retroarch+0x24499c)
    #4 GridView::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) obj-unix/debug/ui/drivers/qt/moc_gridview.cpp:170 (retroarch+0x5fc1b1)
    #5 QMetaObject::activate(QObject*, int, int, void**) <null> (libQt5Core.so.5+0x29ba7b)
    #6 MainWindow::initContentTableWidget() ui/drivers/qt/ui_qt_window.cpp:2659 (retroarch+0x2285b9)
    #7 MainWindow::onCurrentListItemChanged(QListWidgetItem*, QListWidgetItem*) ui/drivers/qt/ui_qt_window.cpp:2391 (retroarch+0x226a1f)
    #8 MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) obj-unix/debug/ui/drivers/moc_ui_qt.cpp:1587 (retroarch+0x5f6aae)
    #9 QMetaObject::activate(QObject*, int, int, void**) <null> (libQt5Core.so.5+0x29ba7b)
    #10 ui_application_qt_run ui/drivers/qt/ui_qt_application.cpp:156 (retroarch+0x212db5)
    #11 rarch_main frontend/frontend.c:157 (retroarch+0x643b9)
    #12 main ui/drivers/qt/ui_qt_application.cpp:182 (retroarch+0x212e8e)

  Previous write of size 8 at 0x7ba000058198 by thread T3:
    #0 memmove /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:720 (libtsan.so.0+0x3295e)
    #1 QCoreApplication::postEvent(QObject*, QEvent*, int) <null> (libQt5Core.so.5+0x275498)

  Location is heap block of size 16384 at 0x7ba000058000 allocated by main thread:
    #0 malloc /build/gcc/src/gcc/libsanitizer/tsan/tsan_interceptors.cc:606 (libtsan.so.0+0x2af84)
    #1 QArrayData::allocate(unsigned long, unsigned long, unsigned long, QFlags<QArrayData::AllocationOption>) <null> (libQt5Core.so.5+0xc4a23)
    #2 ui_companion_driver_toggle ui/ui_companion_driver.c:176 (retroarch+0x65afe)
    #3 ui_companion_driver_init_first ui/ui_companion_driver.c:152 (retroarch+0x65930)
    #4 rarch_main frontend/frontend.c:135 (retroarch+0x6436b)
    #5 main ui/drivers/qt/ui_qt_application.cpp:182 (retroarch+0x212e8e)

  Thread T3 'QXcbEventReader' (tid=16450, running) created by main thread at:
    #0 pthread_create /build/gcc/src/gcc/libsanitizer/tsan/tsan_interceptors.cc:915 (libtsan.so.0+0x2bf03)
    #1 QThread::start(QThread::Priority) <null> (libQt5Core.so.5+0xc38b1)
    #2 ui_companion_qt_init ui/drivers/ui_qt.cpp:271 (retroarch+0x20b3b8)
    #3 ui_companion_driver_init_first ui/ui_companion_driver.c:140 (retroarch+0x65861)
    #4 rarch_main frontend/frontend.c:135 (retroarch+0x6436b)
    #5 main ui/drivers/qt/ui_qt_application.cpp:182 (retroarch+0x212e8e)

SUMMARY: ThreadSanitizer: data race (/usr/lib/libQt5Core.so.5+0x275498) in QCoreApplication::postEvent(QObject*, QEvent*, int)
==================
WARNING: ThreadSanitizer: data race (pid=16366)
  Read of size 8 at 0x7b500021fe18 by thread T14:
    #0 memcpy /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:737 (libtsan.so.0+0x32ace)
    #1 <null> <null> (libQt5Gui.so.5+0x17e99d)
    #2 QtConcurrent::VoidStoredMemberFunctionPointerCall2<void, PlaylistModel, QModelIndex const&, QModelIndex, QString const&, QString>::runFunctor() /usr/include/qt/QtConcurrent/qtconcurrentstoredfunctioncall.h:929 (retroarch+0x279ac2)
    #3 QtConcurrent::RunFunctionTask<void>::run() /usr/include/qt/QtConcurrent/qtconcurrentrunbase.h:136 (retroarch+0x27321c)
    #4 <null> <null> (libQt5Core.so.5+0xbbe10)

  Previous write of size 8 at 0x7b500021fe18 by main thread:
    [failed to restore the stack]

  Location is heap block of size 512 at 0x7b500021fe00 allocated by main thread:
    #0 malloc /build/gcc/src/gcc/libsanitizer/tsan/tsan_interceptors.cc:606 (libtsan.so.0+0x2af84)
    #1 QArrayData::allocate(unsigned long, unsigned long, unsigned long, QFlags<QArrayData::AllocationOption>) <null> (libQt5Core.so.5+0xc4a23)
    #2 operator+(QString const&, char const*) /usr/include/qt/QtCore/qstring.h:1352 (retroarch+0x210318)
    #3 PlaylistModel::getThumbnailPath(QModelIndex const&, QString) const ui/drivers/qt/playlist.cpp:155 (retroarch+0x267f8f)
    #4 PlaylistModel::getCurrentTypeThumbnailPath(QModelIndex const&) const ui/drivers/qt/playlist.cpp:164 (retroarch+0x268265)
    #5 PlaylistModel::loadThumbnail(QModelIndex const&) ui/drivers/qt/playlist.cpp:213 (retroarch+0x26879e)
    #6 MainWindow::updateVisibleItems() ui/drivers/qt/ui_qt_window.cpp:660 (retroarch+0x219df5)
    #7 MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) obj-unix/debug/ui/drivers/moc_ui_qt.cpp:1640 (retroarch+0x5f7442)
    #8 QMetaObject::activate(QObject*, int, int, void**) <null> (libQt5Core.so.5+0x29ba7b)
    #9 ui_application_qt_run ui/drivers/qt/ui_qt_application.cpp:156 (retroarch+0x212db5)
    #10 rarch_main frontend/frontend.c:157 (retroarch+0x643b9)
    #11 main ui/drivers/qt/ui_qt_application.cpp:182 (retroarch+0x212e8e)

  Thread T14 'Thread (pooled)' (tid=16533, running) created by main thread at:
    #0 pthread_create /build/gcc/src/gcc/libsanitizer/tsan/tsan_interceptors.cc:915 (libtsan.so.0+0x2bf03)
    #1 QThread::start(QThread::Priority) <null> (libQt5Core.so.5+0xc38b1)
    #2 QtConcurrent::RunFunctionTaskBase<void>::start() /usr/include/qt/QtConcurrent/qtconcurrentrunbase.h:78 (retroarch+0x276080)
    #3 QFuture<void> QtConcurrent::run<void, PlaylistModel, QModelIndex const&, QModelIndex, QString const&, QString>(PlaylistModel*, void (PlaylistModel::*)(QModelIndex const&, QString const&), QModelIndex const&, QString const&) /usr/include/qt/QtConcurrent/qtconcurrentrun.h:290 (retroarch+0x273f58)
    #4 PlaylistModel::loadThumbnail(QModelIndex const&) ui/drivers/qt/playlist.cpp:218 (retroarch+0x268839)
    #5 MainWindow::updateVisibleItems() ui/drivers/qt/ui_qt_window.cpp:660 (retroarch+0x219df5)
    #6 MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) obj-unix/debug/ui/drivers/moc_ui_qt.cpp:1640 (retroarch+0x5f7442)
    #7 QMetaObject::activate(QObject*, int, int, void**) <null> (libQt5Core.so.5+0x29ba7b)
    #8 ui_application_qt_run ui/drivers/qt/ui_qt_application.cpp:156 (retroarch+0x212db5)
    #9 rarch_main frontend/frontend.c:157 (retroarch+0x643b9)
    #10 main ui/drivers/qt/ui_qt_application.cpp:182 (retroarch+0x212e8e)

SUMMARY: ThreadSanitizer: data race (/usr/lib/libQt5Gui.so.5+0x17e99d) 
==================
WARNING: ThreadSanitizer: data race (pid=16366)
  Write of size 8 at 0x7b740002da00 by thread T12:
    #0 operator delete(void*) /build/gcc/src/gcc/libsanitizer/tsan/tsan_new_delete.cc:74 (libtsan.so.0+0x76004)
    #1 <null> <null> (libQt5Core.so.5+0xbbf21)

  Previous write of size 8 at 0x7b740002da00 by main thread:
    [failed to restore the stack]

  Thread T12 'Thread (pooled)' (tid=16531, running) created by main thread at:
    #0 pthread_create /build/gcc/src/gcc/libsanitizer/tsan/tsan_interceptors.cc:915 (libtsan.so.0+0x2bf03)
    #1 QThread::start(QThread::Priority) <null> (libQt5Core.so.5+0xc38b1)
    #2 QtConcurrent::RunFunctionTaskBase<void>::start() /usr/include/qt/QtConcurrent/qtconcurrentrunbase.h:78 (retroarch+0x276080)
    #3 QFuture<void> QtConcurrent::run<void, PlaylistModel, QModelIndex const&, QModelIndex, QString const&, QString>(PlaylistModel*, void (PlaylistModel::*)(QModelIndex const&, QString const&), QModelIndex const&, QString const&) /usr/include/qt/QtConcurrent/qtconcurrentrun.h:290 (retroarch+0x273f58)
    #4 PlaylistModel::loadThumbnail(QModelIndex const&) ui/drivers/qt/playlist.cpp:218 (retroarch+0x268839)
    #5 MainWindow::updateVisibleItems() ui/drivers/qt/ui_qt_window.cpp:660 (retroarch+0x219df5)
    #6 MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) obj-unix/debug/ui/drivers/moc_ui_qt.cpp:1640 (retroarch+0x5f7442)
    #7 QMetaObject::activate(QObject*, int, int, void**) <null> (libQt5Core.so.5+0x29ba7b)
    #8 ui_application_qt_run ui/drivers/qt/ui_qt_application.cpp:156 (retroarch+0x212db5)
    #9 rarch_main frontend/frontend.c:157 (retroarch+0x643b9)
    #10 main ui/drivers/qt/ui_qt_application.cpp:182 (retroarch+0x212e8e)

SUMMARY: ThreadSanitizer: data race /build/gcc/src/gcc/libsanitizer/tsan/tsan_new_delete.cc:74 in operator delete(void*)
```

Overall it looks pretty good though. @twinaphex might want the variable declarations at the top of the functions though, like the rest of the code (we try to follow C89 everywhere as closely as possible).

CozmoP commented:
> 
> 
> Testing it for now, pretty cool and fast!
> I'm just noticing a crash in Grid view when having a search active and switching playlists.
> It's a bit random, seems to happen faster when searching with at least 3 letters.
> 
> edit: It's crashing even without searching today, just switching playlists (selecting a playlist and pushing down arrow to go through them fast in succession).
> 
> win7 x64, compiled with MSYS2 GCC7.3

Please test with the new commits and see if it's fixed.

CozmoP commented:
> 
> 
> One issue I did notice is that the row numbers in the list view are now out of order, and clicking the name column to sort ascending/descending no longer works. Can you take a look at that?

I've re-enabled sorting by clicking on the column header.
The numbers were showing the position the content was actually in the playlist (the .lpl file).
I've just hidden them for now, but I can make them show in order. Don't really see the point in having them. Maybe we could show the total count somewhere (statusbar or next to the playlist name?)

CozmoP commented:
> 
> 
> The item selection highlight sometimes stop showing in grid view.
> The blue highlight in that pic:
> ![untitled-1](https://user-images.githubusercontent.com/6266038/49903865-cb8efe00-fe68-11e8-9a72-086e9f17fe09.png)
> 
> It happens in long playlists after moving back and forth for a while.
> There was a point where I couldn't select an item or launch it at all.

I can't reproduce. What do you mean by moving back and forth? What Qt version?

CozmoP commented:
> 
> 
> I also get a bunch of TSAN warnings but not sure if they are false positive or not:
> 
> ```
> ...
> ```
> 
I don't know...
> Overall it looks pretty good though. @twinaphex might want the variable declarations at the top of the functions though, like the rest of the code (we try to follow C89 everywhere as closely as possible).

I tried declaring variables at the top of their scope, but I'll take a look.

Tatsuya79 commented:
No more crashing when switching playlists in grid view. Thanks!

I can't reproduce the "highlight disappearing/impossible to select an item" bug any more, just tried for a while. Well, I'll bring that back if it happens again and I can find what was triggering it.

Some suggestions, but could be for later:

-the grid thumbnail type (cover/snap/title) would be better in the footer (where zoom and view are, could be inside "view" or another button) for quick access

-the spacing between columns can change now in grid view when zooming, it was regular before, leaving a spacing on the right side, I think it looked cleaner 
(unless it was the reason for [the search results to be scrambled?](https://github.com/libretro/RetroArch/issues/7143) because this issue is gone with your new code)

CozmoP commented:
> 
> 
> No more crashing when switching playlists in grid view. Thanks!
> 
> I can't reproduce the "highlight disappearing/impossible to select an item" bug any more, just tried for a while. Well, I'll bring that back if it happens again and I can find what was triggering it.
> 
Glad to hear.
> Some suggestions, but could be for later:
> 
> -the grid thumbnail type (cover/snap/title) selector would be better in the footer (where zoom and view are, could be inside "view" or another button) for quick access
> 
Yeah, having the combo there would be useful.
> -the spacing between columns can change now in grid view when zooming, it was regular before, leaving a spacing on the right side, I think it looked cleaner
>
Well I think it looks better now :). It's not when zooming, but when resizing the window/grid.
Try it with the default theme, which I didn't update to keep the look from before so the items don't have a background when not selected.
I actually made 3 different layouts, but didn't expose them in the UI.
In gridview.h, change 
`ViewMode m_viewMode = Centered;`
to
`ViewMode m_viewMode = Simple;`
(and compile)
And there is a third one, "Anchored", which I copied from Windows File Explorer: the first column is fixed and the rest expand to fill the available space.
Maybe make this configurable? I'm not sure on the naming.
> (unless it was the reason for [the search results to be scrambled?](https://github.com/libretro/RetroArch/issues/7143) because this issue is gone with your new code)
>
The reason was that the spacing for the hidden items was still being drawn, so if the search matched two consecutive items, there was no extra space, but if there were items inbetween, they would be separated by all the spacing that would be between as many items. I hope I'm making sense.

Tatsuya79 commented:
Just noticed the history playlist is ordered now instead of putting the most recently launched item on top.

I'm not too bothered by the numbers being displayed or not in the playlist.
I would show the total number of items as you suggested.
I think the best place for that is into the footer we have in grid view (with "zoom" and "view", on the opposite side, to the left).

This footer would be displayed in places where it is not atm: detailed view and file browser.
It would adapt its content: "Total: xx items" everywhere, "view" for grid and detailed list, file browser could also have "total size: xxx KB" or individual selected item size...

Tatsuya79 commented:
That works, but you've got several other history playlists:
"image", "music", "video".

bparker06 commented on ui/drivers/qt/gridview.cpp:
```diff
@@ -332,7 +339,9 @@ void GridView::setSelection(const QRect &rect, QFlags<QItemSelectionModel::Selec
+
+   foreach(range, selection)
    {
       for (int row = range.top(); row <= range.bottom(); ++row)
```
the int declaration should come before the for loop

bparker06 commented on ui/drivers/qt/gridview.cpp:
```diff
@@ -332,7 +339,9 @@ void GridView::setSelection(const QRect &rect, QFlags<QItemSelectionModel::Selec
-   foreach(const QItemSelectionRange &range, selection)
+   QItemSelectionRange range;
+
+   foreach(range, selection)
```
Just FYI for in the future, we try to avoid `foreach` because it makes a copy of all the items, and it is scheduled for deprecation. See [here](https://www.kdab.com/goodbye-q_foreach/) and [here](https://codereview.qt-project.org/#/c/147363/).

bparker06 commented on ui/drivers/qt/ui_qt_window.cpp:
```diff
@@ -277,13 +247,14 @@ MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent)
    ,m_loadCoreWindow(new LoadCoreWindow(this))
    ,m_timer(new QTimer(this))
+   ,m_thumbnailTimer(new QTimer(this))
```
This needs to go at the bottom of the init list, as in my diff here that I provided earlier:
```c
diff --git a/ui/drivers/qt/ui_qt_window.cpp b/ui/drivers/qt/ui_qt_window.cpp
index ccec679b9d..579a7142b6 100644
--- a/ui/drivers/qt/ui_qt_window.cpp
+++ b/ui/drivers/qt/ui_qt_window.cpp
@@ -247,7 +247,6 @@ MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent)
    ,m_loadCoreWindow(new LoadCoreWindow(this))
    ,m_timer(new QTimer(this))
-   ,m_thumbnailTimer(new QTimer(this))
    ,m_currentCore()
    ,m_currentCoreVersion()
    ,m_statusLabel(new QLabel(this))
@@ -322,6 +321,7 @@ MainWindow::MainWindow(QWidget *parent) :
    ,m_failedThumbnails(0)
    ,m_playlistThumbnailDownloadWasCanceled(false)
    ,m_pendingDirScrollPath()
+   ,m_thumbnailTimer(new QTimer(this))
 {
    settings_t *settings = config_get_ptr();
    QDir playlistDir(settings->paths.directory_playlist);
```

twinaphex replied:
Hey @CozmoP , if you can address these last few issues that @bparker06 pointed out, I think we can start fast-tracking this for merging.

twinaphex replied:
Hey @CozmoP , another request to address these issues by @bparker06 , then we can finally merge this.

CozmoP replied:
Hi @twinaphex , I believe I addressed these issues, but see my last comment.

Tatsuya79 commented:
I forgot to report that one: thumbnails aren't centred vertically in their grid space.
![untitled-1](https://user-images.githubusercontent.com/6266038/50119689-c7415700-0253-11e9-95e3-86090a11ad99.jpg)

bparker06 commented on ui/drivers/qt/gridview.cpp:
```diff
@@ -340,12 +342,16 @@ QRegion GridView::visualRegionForSelection(const QItemSelection &selection) cons
    {
-      for (int row = range.top(); row <= range.bottom(); ++row)
+      range = selection.at(i);
+      int row = range.top();
```
doh

CozmoP commented:
> That works, but you've got several other history playlists:
> "image", "music", "video".

I think the current version is sorting those aswell, so I'll leave it for now.

> I forgot to report that one: thumbnails aren't centred vertically in their grid space.

I also did that on purpose :P. So I made it possible to change the style of the grid view and items from the style sheet. The best way i found is adding custom properties, so for now I added:

for selector **GridView**

Property | Type | Description
-------- | ---- | --------
qproperty-spacing | Integer | (Minimum) space between items and between items and the edge of the viewport. Default: 7.
qproperty-layout | String | The layout of the grid. Possible values are:<br>`"fixed"`: items are separated by exactly `qproperty-spacing`, starting from the left, until there is no more space to fit a new column, in which case there will be empty space.<br>`"centered"`: items are separated to fill the available space, with a minimum spacing of `qproperty-spacing`.<br>`"anchored"`: first column from the left is separated from the viewport by `qproperty-spacing`. The rest are separated to fill the available space, with a minimum spacing of `qproperty-spacing`.<br>Default: `"centered"`.

for selector **GridItem**

Property | Type | Description
-------- | ---- | --------
qproperty-padding | Integer | Space from the edge of the item and its contents, and between the text and the image. Default: 11.
qproperty-valign | String | Vertical alignment of the thumbnail image. Possible values are `"top"`, `"center"` and `"bottom"`. Default: `"bottom"`.

I updated the dark theme so it keeps the same look as before my changes.
The default theme is using the defaults, which I set as I think they should be, but feel free to change them.

CozmoP commented:
I noticed another issue: when on the file browser, the file extensions don't show, and you can edit the entries (although nothing is really edited other than the text you see).
So while fixing that I thought of making a different table altogether and make it like an actual file browser, like this:

![imagen](https://user-images.githubusercontent.com/25121396/50666990-88333580-0fb7-11e9-8682-7b333d36489d.png)

I pushed this to another branch since it's out of the scope of this PR, but if you want I can merge it to this one.
Please have a look at https://github.com/CozmoP/RetroArch/tree/filebrowser
(It still shows the thumbnails).

Tatsuya79 commented:
I like the file browser view with more details.

I just noticed some things while testing it though (just writing them here to remember):
-the default "start directory" browser view displays the different drives instead of the folder content
-when into a rom folders with thumbnails, clicking with the mouse on an entry refreshes the thumbs while moving with up/down arrow keys does not
-the file names are short, it would be better to save the column size globally for the detailed view

CozmoP commented:
>
> -the file names are short, it would be better to save the column size globally for the detailed view

Done.

Tatsuya79 commented:
I still have the filename column changing its size when clicking on different folders.
I see the files first, then, if sub-folders are present, the list gets refreshed with them moved at the top.
Is it trying to save the size for every different folder?
